### PR TITLE
Feature/farm polling refactor v2

### PR DESCRIPTION
**Pull request by hachiojidev**
- Created a global farm data fetching hook that only fetches farm data
- Refactored the farm fetching thunk to accept an array of `pids` so different hooks can fetch specific farms. This should allow the ability to add in "smart" fetching tied to our lazy loading.
- Refactored staking hooks to make them more pure. Now the consumers handle the refreshing of the data.

### CODE REVIEW

File: src/hooks/useApprove.ts [hunk 1]
```diff
@@ -3,25 +3,23 @@ import { useWeb3React } from '@web3-react/core'
...
 // Approve a Farm
 export const useApprove = (lpContract: Contract) => {
-  const dispatch = useAppDispatch()
   const { account } = useWeb3React()
   const masterChefContract = useMasterchef()
 
   const handleApprove = useCallback(async () => {
     try {
       const tx = await approve(lpContract, masterChefContract, account)
-      dispatch(fetchFarmUserDataAsync(account))
```
Comment (hachiojidev) [inline src/hooks/useApprove.ts:19]: I felt this was a side-effect that should not be in this hook. Instead I moved the re-fetching of the user data to the places that use this hook.

File: src/hooks/useStake.ts [hunk 2]
```diff
@@ -1,22 +1,20 @@
...
 
 const useStake = (pid: number) => {
-  const dispatch = useAppDispatch()
   const { account } = useWeb3React()
   const masterChefContract = useMasterchef()
 
   const handleStake = useCallback(
     async (amount: string) => {
       const txHash = await stake(masterChefContract, pid, amount, account)
-      dispatch(fetchFarmUserDataAsync(account))
```
Comment (hachiojidev) [inline src/hooks/useStake.ts:16]: Same here, moved it to the consumers that use this handler.

File: src/hooks/useHarvest.ts [hunk 3]
```diff
@@ -12,7 +12,7 @@ export const useHarvest = (farmPid: number) => {
 
   const handleHarvest = useCallback(async () => {
     const txHash = await harvest(masterChefContract, farmPid, account)
-    dispatch(fetchFarmUserDataAsync(account))
+    dispatch(fetchFarmUserDataAsync({ account, pids: [farmPid] }))
```
Comment (hachiojidev) [inline src/hooks/useHarvest.ts:15]: This handler now only updates the relevant farm.

File: src/hooks/useUnstake.ts [hunk 4]
```diff
@@ -1,27 +1,20 @@
...
 
 const useUnstake = (pid: number) => {
-  const dispatch = useAppDispatch()
   const { account } = useWeb3React()
   const masterChefContract = useMasterchef()
 
   const handleUnstake = useCallback(
     async (amount: string) => {
       const txHash = await unstake(masterChefContract, pid, amount, account)
-      dispatch(fetchFarmUserDataAsync(account))
```
Comment (hachiojidev) [inline src/hooks/useUnstake.ts:21]: And...finally moved this one to the consumer as well.

File: src/state/farms/index.ts [hunk 5]
```diff
@@ -27,68 +24,87 @@ const noAccountFarmConfig = farmsConfig.map((farm) => ({
...
+
+    const farms = await fetchFarms(farmsWithPriceHelpers)
+    const farmsWithPrices = await fetchFarmsPrices(farms)
+
+    // Filter out price helper LP config farms
+    const farmsWithoutHelperLps = farmsWithPrices.filter((farm: Farm) => {
+      return farm.pid || farm.pid === 0
+    })
+
+    return farmsWithoutHelperLps
```
Comment (hachiojidev) [inline src/state/farms/index.ts:44]: Same logic as the old thunk

File: src/state/farms/index.ts [hunk 6]
```diff
@@ -27,68 +24,87 @@ const noAccountFarmConfig = farmsConfig.map((farm) => ({
...
+    return userFarmAllowances.map((farmAllowance, index) => {
+      return {
+        pid: farmsToFetch[index].pid,
+        allowance: userFarmAllowances[index],
+        tokenBalance: userFarmTokenBalances[index],
+        stakedBalance: userStakedBalances[index],
+        earnings: userFarmEarnings[index],
+      }
+    })
+  },
```
Comment (hachiojidev) [inline src/state/farms/index.ts:74]: No logic change.

File: src/state/hooks.ts [hunk 7]
```diff
@@ -111,7 +138,7 @@ export const useLpTokenPrice = (symbol: string) => {
     // Double it to get overall value in LP
     const overallValueOfAllTokensInFarm = valueOfBaseTokenInFarm.times(2)
     // Divide total value of all tokens, by the number of LP tokens
-    const totalLpTokens = getBalanceAmount(farm.lpTotalSupply)
+    const totalLpTokens = getBalanceAmount(new BigNumber(farm.lpTotalSupply))
```
Comment (hachiojidev) [inline src/state/hooks.ts:141]: Due to updating the Farm type.

File: src/views/Farms/components/FarmCard/StakeAction.tsx [hunk 8]
```diff
@@ -36,6 +39,18 @@ const StakeAction: React.FC<FarmCardActionsProps> = ({
   const { onStake } = useStake(pid)
   const { onUnstake } = useUnstake(pid)
   const location = useLocation()
+  const dispatch = useAppDispatch()
+  const { account } = useWeb3React()
+
+  const handleStake = async (amount: string) => {
+    await onStake(amount)
+    dispatch(fetchFarmUserDataAsync({ account, pids: [pid] }))
```
Comment (hachiojidev) [inline src/views/Farms/components/FarmCard/StakeAction.tsx:47]: This is where we refresh the user data (originally it was in `onStake` and was refreshing all user data)

File: src/views/Pools/index.tsx [hunk 9]
```diff
@@ -44,6 +44,7 @@ const Pools: React.FC = () => {
 
   useFetchCakeVault()
   useFetchPublicPoolsData()
+  usePollFarmsData()
```
Comment (hachiojidev) [inline src/views/Pools/index.tsx:47]: Since the pools need the farms we have to live with fetching all the farms data here as well.

**Comment by hachiojidev** [comment]
Did a quick test on some pages. Connected wallet, refreshed the page and waited for 1 minute. Recorded the total XHR calls.

| Page        | Pre Pools/Farms refactor) | New   | Change | Notes                                                                    |
| ----------- | ------------------------- | ----- | ------ | ------------------------------------------------------------------------ |
| Home        | 1,262                     | 689   | -45%   | Makes a one-time call to fetch farm data for highest APR                 |
| Farms       | 1,243                     | 1,222 | -1.7%  | Same as before, plus additional core farm calls                          |
| Pools       | 1,214                     | 1,354 | +11.5% | Fetching farms data plus pools data                                      |
| Predictions | 1,324                     | 227   | -82.9% | No more farm or pool calls. Only predictions and global "core" farm data |
| IFO         | 1,218                     | 131   | -89.2% | No more farm or pool calls. Only IFO and global "core" farm data         |


Any page that does not require farm or pool data will see 80%+ reduction in calls.

File: src/views/Home/Home.tsx [hunk 10]
```diff
@@ -86,6 +88,12 @@ const CTACards = styled(BaseLayout)`
 
 const Home: React.FC = () => {
   const { t } = useTranslation()
+  const dispatch = useAppDispatch()
+
+  // Fetch farm data once to get the max APR
+  useEffect(() => {
+    dispatch(fetchFarmsPublicDataAsync(nonArchivedFarms.map((nonArchivedFarm) => nonArchivedFarm.pid)))
```
Comment (Chef-Cheems) [inline src/views/Home/Home.tsx:95]: Could further decrease amount of calls if we'd fetch only LIVE farms. NonArchicved includes finished too which is a bit useless.

File: src/hooks/useHarvest.ts (hunk 3 above)
Comment (Chef-Cheems) [reply src/hooks/useHarvest.ts:15]: Should we move it to the consumer side as with other hooks?

File: src/views/Home/Home.tsx (hunk 10 above)
Comment (hachiojidev) [reply src/views/Home/Home.tsx:95]: Ah, I didn't even think of that. Will update thanks!

**Review by Chef-Cheems — APPROVED** [review]
Looks good!
There is probably more room for decreasing amount of requests but for upcoming IFO this should help tremendously :+1:

File: src/state/types.ts [hunk 11]
```diff
@@ -11,14 +11,14 @@ export type TranslatableText =
...
 export interface Farm extends FarmConfig {
-  tokenAmountMc?: BigNumber
-  quoteTokenAmountMc?: BigNumber
-  tokenAmountTotal?: BigNumber
-  quoteTokenAmountTotal?: BigNumber
-  lpTotalInQuoteToken?: BigNumber
-  lpTotalSupply?: BigNumber
-  tokenPriceVsQuote?: BigNumber
-  poolWeight?: BigNumber
+  tokenAmountMc?: string
```
Comment (ChefHutch) [inline src/state/types.ts:14]: You love to see it 🙌

File: src/state/types.ts [hunk 12]
```diff
@@ -10,15 +10,17 @@ export type TranslatableText =
...
 export interface Farm extends FarmConfig {
-  tokenAmountMc?: BigNumber
-  quoteTokenAmountMc?: BigNumber
-  tokenAmountTotal?: BigNumber
-  quoteTokenAmountTotal?: BigNumber
-  lpTotalInQuoteToken?: BigNumber
-  lpTotalSupply?: BigNumber
-  tokenPriceVsQuote?: BigNumber
-  poolWeight?: BigNumber
+  tokenAmountMc?: SerializedBigNumber
```
Comment (ChefHutch) [inline src/state/types.ts:16]: You love to see it... even more 🙌 🥳

File: src/views/Pools/index.tsx (hunk 9 above)
Comment (ChefHutch) [reply src/views/Pools/index.tsx:47]: Yea makes sense. It's just for prices right? One to think about in future.

**Review by RabbitDoge — APPROVED** [review]

**Review by ChefHutch — APPROVED** [review]

**Pull request merged**